api/src/controllers/payment/payments.ts

```diff
@@ -0,0 +1,375 @@
+  ordersQueries,
+  paymentLogQueries,
+  userSubscriptionQueries,
+  packagesQueries,
+} from "@/graphql/queries";
+import {
+  ordersMutations,
+  paymentLogMutations,
+  userSubscritpionMutation,
+} from "@/graphql/mutations";
+
+// Initialize VNPay
+// https://vnpay.js.org/create-payment-url
+const vnpay = new VNPay({
+  tmnCode: config.VNPAY_TMNCODE,
+  secureSecret: config.VNPAY_SECRET_KEY,
+  vnpayHost: "https://sandbox.vnpayment.vn",
+  testMode: true,
+  hashAlgorithm: HashAlgorithm.SHA512,
+
+  enableLog: true,
+});
+
+const setOrderStatus = async (orderId: string, status: string) => {
+  const statuses = ["Pending", "Paid", "Failed", "Chargeback", "Refunded"];
+
+  if (!statuses.includes(status)) {
+    throw new Error("Invalid status!");
+  }
+
+  const order = await graphQLClient().request(
+    ordersMutations.UPDATE_ORDER_STATUS,
+    {
+      id: orderId,
+      status: status,
+    }
+  );
+
+  return order;
+};
+
+const createPaymentLog = async (
+  userId: string,
+  orderId: string,
+  request: string,
+  response: string,
+  transactionId: string,
+  logType: string
+) => {
+  if (!userId) {
+    throw new Error("User ID is required");
+  }
+
+  if (!orderId) {
+    throw new Error("Order ID is required");
+  }
+
+  const logTypes = ["authorize", "redirect", "IPN"];
+  if (!logTypes.includes(logType)) {
+    throw new Error("Invalid log type");
+  }
+
+  const paymentLog = await graphQLClient().request(
+    paymentLogMutations.CREATE_PAYMENT_LOG,
+    {
+      user_id: userId,
+      order_id: orderId,
+      request: request,
+      response: response,
+      transaction_id: transactionId,
+      log_type: logType,
+    }
+  );
+
+  return paymentLog;
+};
+
+// ===================
+
+export const createOrder = async (
+  req: express.Request,
+  res: express.Response,
+  next: express.NextFunction
+) => {
+  try {
+    let ipAddress =
+      req.headers["x-forwarded-for"] ||
+      req.connection.remoteAddress ||
+      req.socket.remoteAddress ||
+      req.ip ||
+      "";
+
+    if (typeof ipAddress !== "string") {
+      ipAddress = ipAddress[0];
+    }
+
+    // Get the input
+    const uid = req.body.uid ?? res.locals.uid;
+    const packageId = req.body.packageId;
+    const bankCode = req.body.bankCode;
+    let orderInfo = req.body.orderDescription ?? "";
+    const locale = req.body.locale ?? "vn";
+
+    // Get the package
+    if (!packageId) {
+      res.status(400).json({
+        message: "Package ID is required",
+        success: false,
+      });
+
+      return;
+    }
+
+    const packageData = await graphQLClient().request(
+      packagesQueries.GET_PACKAGE,
+      {
+        package_id: packageId,
+      }
+    );
+
+    if (packageData.package === null) {
+      res.status(404).json({
+        message: "Package not found",
+        success: false,
+      });
+
+      return;
+    }
+
+    // Create an Order document
+    const amount = Number(req.body.amount) ?? packageData.package.base_price;
```

Comment:
You shouldn't let the API endpoint have the amount as input as the user client could easily adjust the amount to whatever they want. But for the project to be delivered on time (26/12) and this project is within the university course scope, this could be temporarily acceptable.